addons/payment_flutterwave/models/payment_transaction.py

```diff
@@ -0,0 +1,199 @@
+
+        first_name, last_name = payment_utils.split_partner_name(self.partner_name)
+        data = {
+            'token': self.token_id.acquirer_ref,
+            'email': self.token_id.flutterwave_customer_email,
+            'amount': self.amount,
+            'currency': self.currency_id.name,
+            'country': self.company_id.country_id.code,
+            'tx_ref': self.reference,
+            'first_name': first_name,
+            'last_name': last_name,
+            'ip': payment_utils.get_customer_ip_address(),
+        }
+
+        # Make the payment request to Flutterwave.
+        response_content = self.acquirer_id._flutterwave_make_request(
+            'tokenized-charges', payload=data
+        )
+
+        # Handle the payment request response.
+        _logger.info(
+            "payment request response for transaction with reference %s:\n%s",
+            self.reference, pprint.pformat(response_content)
+        )
+        self._handle_notification_data('flutterwave', response_content['data'])
+
+    def _get_tx_from_notification_data(self, provider, notification_data):
+        """ Override of payment to find the transaction based on Flutterwave data.
+
+        :param str provider: The provider of the acquirer that handled the transaction.
+        :param dict notification_data: The notification data sent by the provider.
+        :return: The transaction if found.
+        :rtype: recordset of `payment.transaction`
+        :raise ValidationError: If inconsistent data were received.
+        :raise ValidationError: If the data match no transaction.
+        """
+        tx = super()._get_tx_from_notification_data(provider, notification_data)
+        if provider != 'flutterwave' or len(tx) == 1:
+            return tx
+
+        reference = notification_data.get('tx_ref')
+        if not reference:
+            raise ValidationError("Flutterwave: " + _("Received data with missing reference."))
+
+        tx = self.search([('reference', '=', reference), ('provider', '=', 'flutterwave')])
+        if not tx:
+            raise ValidationError(
+                "Flutterwave: " + _("No transaction found matching reference %s.", reference)
+            )
+        return tx
+
+    def _process_notification_data(self, notification_data):
+        """ Override of payment to process the transaction based on Flutterwave data.
+
+        Note: self.ensure_one()
+
+        :param dict notification_data: The notification data sent by the provider.
+        :return: None
+        :raise ValidationError: If inconsistent data were received.
+        """
+        super()._process_notification_data(notification_data)
+        if self.provider != 'flutterwave':
+            return
+
+        # Verify the notification data.
+        verification_response_content = self.acquirer_id._flutterwave_make_request(
+            'transactions/verify_by_reference', payload={'tx_ref': self.reference}, method='GET'
```

Comment:
The payload received in `_flutterwave_make_request()` is given in the body as a JSON object. However, according to Flutterwave documentation, `tx_ref` should be passed as a query parameter. This causes a bad request error at the return of the Flutterwave payment page.